### ISSUE TITLE
Could you explain why yield* works here and not yield?

### ISSUE THREAD

**Issue by danpantry** [yelouafi/redux-saga#285]
This is probably going to be a RTFM with regards to generators. Here's an excerpt.

```javascript
import { take, fork, put, call } from 'redux-saga/effects'

function* fetchCompanySaga() {
  yield* call(wait, 1000)
  const action = yield take('FETCH_COMPANY')
  yield put({
    type: 'FETCH_COMPANY_SUCCESS',
    payload: {
      id: action.id,
      name: 'Foo'
    }
  })
}

// This could theoretically be any promise such as fetch(), it is just an example.
function wait(ms) {
  return new Promise(function (resolve) {
    setTimeout(() => resolve(), ms)
  })
}

export default function* root() {
  yield fork(fetchCompanySaga)
}
```

And it is called like so:

```javascript
const sagaMiddleware = createSagaMiddleware()
const store = createStore(reducer, applyMiddleware(sagaMiddleware))
const task = sagaMiddleware.run(root)
const promise = task.done

const rendered = (<Provider store={store}>
  <Handler />
</Provider>)

// Any initial actions, for example, if we know the current route component..
store.dispatch(fetchCompany(1))

promise.then(function() {
  // Now the app can be served over HTTP
  console.log(renderToString(rendered))
})
```

Essentially, the 
idea in this *very limited* sample is simply to simulate a saga that might take a while to load up. For example, a one-time only saga - for example, onboarding on a server side render. I'm interested in just working out when the root saga completes.

In this scenario, I know that I'm only going to run a certain number of actions and then the task will complete. Whether or not this is realistic is another question; I'm just trying to get the hang of `redux-saga`. Now, this code does actually work, however when replacing `yield*` with `yield` in place of `call`, the final `promise` never fulfills itself. This indicates to me that I would have to use `yield*` any time I wanted to use a promise-based resource. Is my understanding correct?

If this is the case, what is the appropriate scenario for server-side rendering (if there is one at all)? From my gathering of another issue on this repo, you should use different sagas for different environments - this seems to make sense

**Comment by fhelwanger** [comment]
I don't know why it works with `yield*`, but you probably should change the following line:

```diff
- const action = take('FETCH_COMPANY')
+ const action = yield take('FETCH_COMPANY')
```

**Comment by danpantry** [comment]
@fhelwanger Thanks. That is actually what I had, but I was reconstructing my original example after modifying it a bit (to experiment with `takeEvery`).

**Comment by yelouafi** [comment]
@danpantry the correct form is `yield`. The saga can't take the dispatched action because it's still blocked on the `call(wait, 1000)` when you dispatch the action. So when it resumes from the delay it remains blocked on the take.

Here is the correct code

```javascript
function* fetchCompanySaga() {
  const action = yield take('FETCH_COMPANY')
  // put the delay after the take
  yield call(wait, 1000)
  yield put({
    type: 'FETCH_COMPANY_SUCCESS',
    payload: {
      id: action.id,
      name: 'Foo'
    }
  })
}
```

**Comment by yelouafi** [comment]
as for `yield*` it's used to delegate to other Iterables object (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/yield*). AFAIK, It shouldn't work with non iterables (like the result of `call` above). It seems the regenerator runtime (used by babel to transpile Generator code) resumes the Generator when you yield* to a non iterable instead of throwing an error

**Comment by rahulbasu710** [comment]
Hey,
I am using a React Universal environment and I am loading some data through static component from the router that loads from server initially.

The problem I am facing is that the saga is called properly on the server-side but unless I dispatch the same action again inside either `componentDidMount` or `componentWillMount` it is not reflecting on the client-side store. Also doesn't show up on the `redux-devtools` as well. So the `yield put` is basically the issue

```
import { call, put, takeEvery, all } from 'redux-saga/effects';
import axios from 'axios';
import jwt from 'jsonwebtoken';
import { port, host } from '../appConfig';

// const getHomeState = state => state.home; const homeState = yield select(getHomeState);

export function* fetchData() {
  try {
    const token = jwt.sign({ exp: Math.floor(Date.now() / 1000) + (60 * 60) }, 'wrongrightwhatever');
    let response = null;
    if (port) {
      const reqURL = `http://${host}:${port}/api/data`;
      response = yield call(axios.get, reqURL, { headers: { authorization: token, 'Content-Type': 'application/json' } });
    } else {
      const reqURL = '/api/data';
      response = yield call(axios.get, reqURL, { headers: { authorization: token, 'Content-Type': 'application/json' } });
    }
    console.log(response.data.msg); // sends the response message and shows up on the console
    yield put({ type: 'FETCHED_JWT', data: response.data.msg });
  } catch (err) {
    console.log(err);
  }
}

export function* watchFetchData() {
  yield takeEvery('FETCH_DATA', fetchData); // Take the 1st page data snapshots, draftdata from draftid
}

export default function* rootSaga() {
  yield all([
    watchFetchData(),
  ]);
}
```

Do I need to run the saga dispatching action for sure on the client & server both because then the middleware router function is called like 3 times which I want to avoid
